src/components/RichTextEditor/Toolbar.tsx

```diff
@@ -62,108 +71,213 @@ const toolbarComponents = (
-                icon={<IconListBullets size={IconSize.Size24} />}
+                icon={<WrapIconStyling icon={<IconListBullets size={IconSize.Size24} />} />}
                 classNames={classNames}
                 styles={styles}
             />
         ),
     };
 };
 
-export const Toolbar: FC<ToolbarProps> = ({ editorId, textStyles, actions = [] }) => {
+type ButtonGroupProps = {
+    index: number;
+    editorId?: string;
+    actions: EditorActions[];
+    textStyles?: TextStyleType[];
+    className: string;
+    onClose: () => void;
+    onLoaded: (index: number, width?: number) => void;
+};
+
+const ButtonGroup: FC<ButtonGroupProps> = ({ index, actions, editorId, textStyles, onLoaded, onClose, className }) => {
+    const ref = useRef<HTMLDivElement | null>(null);
+
     const editor = usePlateEditorRef(editorId);
+
+    useEffect(() => {
+        onLoaded(index, ref.current?.clientWidth);
+        onClose();
+    }, [ref.current?.clientWidth]);
+
+    return (
+        <div
+            ref={ref}
+            key={index}
+            data-test-id={`toolbar-group-${index}`}
+            className={merge([
+                "tw-flex tw-items-center tw-border-l tw-border-line tw-h-12 tw-border-b tw-p-2",
+                className,
+            ])}
+        >
+            {actions.map((action, index) => (
+                <React.Fragment key={index}>{toolbarComponents(editor, editorId, textStyles)[action]}</React.Fragment>
+            ))}
+        </div>
+    );
+};
+
+const DEFAULT_MAX_WIDTH = "100%";
+const BORDER_PX = 6;
+
+export const Toolbar: FC<ToolbarProps> = ({ editorId, textStyles, actions = [] }) => {
     const toolbarActions = actions.length > 0 ? actions : defaultActions;
+    const [buttonGroupsWidths, setButtonGroupsWidths] = useState<{ [index: number]: number | undefined }>({});
+    const toolbarRef = useRef<HTMLDivElement | null>(null);
+    const [toolbarWidth, setToolbarWidth] = useState<number>(0);
+    const [cssStyles, setCssStyles] = useState({ maxWidth: toolbarWidth || DEFAULT_MAX_WIDTH });
+    const [borderClassname, setBorderClassname] = useState("");
+
+    useEffect(() => {
+        if (toolbarWidth <= 2) {
+            return;
+        }
+        let sum = 0;
+        const lengthButtonGroups = Object.entries(buttonGroupsWidths).length;
+        const array: number[][] = [...Array(lengthButtonGroups).keys()].map(() => [0]);
+
+        for (const [, value] of Object.entries(buttonGroupsWidths)) {
+            sum += value || 0;
+
+            const index = Math.floor(sum / toolbarWidth);
+            array[index]?.push(value || 0);
+        }
+        const width = Math.max(...array.map((element) => element.reduce((prev, curr) => prev + curr), 0));
+
+        const rowNumber = array.reduce((prev, curr) => (curr.length > 1 ? prev + 1 : prev + 0), 0);
+
+        if (rowNumber === 1) {
+            setBorderClassname("first:tw-border-l-0 tw-border-b-0 first:tw-border-l-0 tw-border-l");
```

Comment:
That's pretty fancy 🙃. 
But let's see if it's ok for the designers.

But can you at least outsource the contents of the two useEffects to named functions?
So that you can see at least roughly what is happening based on the name.

Reply:
Tried to adapt it. Could you check it again?